### PR TITLE
Async object store support

### LINKED ISSUE #2: Async interface
The Rust crate this project is based on provides an async interface. However, this project provides a synchronous interface. Would it be possible to expose an async interface here (perhaps in addition to the existing synchronous interface)?

Context: In the [Zarr-Python project](https://github.com/zarr-developers/zarr-python/blob/main/v3-roadmap-and-design.md#async-api), we are considering using a this library as one of the store backends. In the next version of Zarr-Python, all stores will be async-only so we would want the backend to expose as much async functionality as possible. 

@kylebarron recently pointed me to https://github.com/PyO3/pyo3/issues/1632#issuecomment-1752582018 which suggests "the next version of pyo3 is likely to have big progress in async handling".

cc @martindurant

### PR DESCRIPTION
**Pull request by kylebarron**
### Change list

- Add async methods for each existing function.
- Remove superfluous vec clone. `PyBytes::new(py, &Bytes::to_vec())` had two clones, I believe, where only one was needed.
- [Automatically convert](https://pyo3.rs/v0.20.2/conversions/tables#returning-rust-values-to-python) to `PyBytes` from `Cow<[u8]>` instead of converting manually. At least for the async version this was a lot easier; I struggled with manually constructing `PyBytes` objects inside the async closure.

Another option for the API would be to have a separate class `AsyncObjectStore` that has only the async methods. This might keep the methods cleaner (we wouldn't need to have an `async` suffix on each method). But alternatively, there's no reason why someone couldn't use the same store for both synchronous and async calls.

This is implemented on top of https://github.com/roeap/object-store-python/pull/5, so ignore the first several commits. It's expected for #5 to be merged first, and then this PR can be squashed.

Closes #2

### CODE REVIEW

File: object-store/src/lib.rs [hunk 1]
```diff
@@ -514,29 +515,82 @@ impl PyObjectStore {
...
             .block_on(self.inner.get_range(&location.into(), range))
-            .map_err(ObjectStoreError::from)?
-            .to_vec();
-        Python::with_gil(|py| Ok(PyBytes::new(py, &obj).into_py(py)))
+            .map_err(ObjectStoreError::from)?;
+        Ok(Cow::Owned(obj.to_vec()))
+    }
+
+    /// Return the bytes that are stored at the specified location in the given byte range
+    #[pyo3(text_signature = "($self, location)")]
```
Comment (jhamman) [inline object-store/src/lib.rs]: Is this missing the start/length parameters

Comment (kylebarron) [reply object-store/src/lib.rs]: Yeah
```suggestion
    #[pyo3(text_signature = "($self, location, start, length)")]
```

**Comment by ion-elgreco** [comment]
This looks interesting! Any chance to get this merged and released? @roeap

**Comment by simonbutt** [comment]
Any progress getting this released?

**Comment by jhamman** [comment]
No progress here. The store API has managed to mature a bit so if someone wanted to come and pick this up (@kylebarron or otherwise), now would be a good time.

**Comment by dark0dave** [comment]
> No progress here. The store API has managed to mature a bit so if someone wanted to come and pick this up (@kylebarron or otherwise), now would be a good time.

I have picked it up.

**Comment by ion-elgreco** [comment]
Maybe it's easier to fork, and then merge all open PRs into that fork and release it

**Review by roeap — APPROVED** [review]
Thanks for taking care of this @kylebarron! And sorry for letting this sit for so long.

**Pull request merged**